taskvine/src/worker/vine_cache.c

```diff
@@ -376,99 +374,160 @@ static int do_transfer( struct vine_cache *c, const char *source_url, const char
 */
 
-int vine_cache_ensure( struct vine_cache *c, const char *cachename, struct link *manager )
+vine_file_status_type_t vine_cache_ensure( struct vine_cache *c, const char *cachename)
 {
-	if(!strcmp(cachename,"0")) return 1;
+	if(!strcmp(cachename,"0")) return VINE_FILE_STATUS_READY;
 
 	struct cache_file *f = hash_table_lookup(c->table,cachename);
 	if(!f) {
 		debug(D_VINE,"cache: %s is unknown, perhaps it failed to transfer earlier?",cachename);
-		return 0;
+		return VINE_FILE_STATUS_FAILED;
 	}
 
-	if(f->complete) {
-		/* File is already present in the cache. */
-		return 1;
+	/* File is already present in the cache. */
+	if(f->status==VINE_FILE_STATUS_READY)  return VINE_FILE_STATUS_READY;
+	/* transfer process completed and failed. */
+	if(f->status==VINE_FILE_STATUS_FAILED) return VINE_FILE_STATUS_FAILED;
+	/* transfer process running. */
+	if(f->status==VINE_FILE_STATUS_PROCESSING) return VINE_FILE_STATUS_PROCESSING;
+
+	if(f->type == VINE_CACHE_MINI_TASK){
+		if(f->mini_task->input_mounts) {
+                	struct vine_mount *m;
+			vine_file_status_type_t result;
+                	LIST_ITERATE(f->mini_task->input_mounts,m) {
+                        	result = vine_cache_ensure(c,m->file->cached_name);
+                        	if(result!=VINE_FILE_STATUS_READY) return result;
+                	}
+        	}
 	}
+	
+	f->start_time = timestamp_get();
+
+	debug(D_VINE,"forking transfer process to create %s", cachename);
+
+	struct vine_process *p;
+	if(f->type == VINE_CACHE_MINI_TASK){
+		p = vine_process_create(f->mini_task, 1);
+		vine_sandbox_stagein(p, c);
+		f->process = p;
+	}
+	pid_t pid = fork();
+
+	if(pid == -1) {
+		debug(D_VINE,"failed to fork transfer process");
+		return VINE_FILE_STATUS_FAILED;
+	}
+	if(pid > 0){
+		f->pid = pid;
+		f->status = VINE_FILE_STATUS_PROCESSING;
+		if(f->type==VINE_CACHE_TRANSFER) debug(D_VINE,"cache: transferring %s to %s",f->source,cachename);
+		if(f->type==VINE_CACHE_MINI_TASK) debug(D_VINE,"cache: creating %s via mini task",cachename);
+		return VINE_FILE_STATUS_PROCESSING;
+	}
+
+	vine_cache_get_file(f, c, cachename);
+	exit(1);
+
+}
+/*
+Child Process that materializes the proper file.
+
+*/
+
+void vine_cache_get_file(struct cache_file *f, struct vine_cache *c, const char *cachename){
 
 	char *error_message = 0;
 	char *cache_path = vine_cache_full_path(c,cachename);
-
 	int result = 0;
 
-	timestamp_t transfer_start = timestamp_get();
-
 	switch(f->type) {
 		case VINE_CACHE_FILE:
-			debug(D_VINE,"cache: manager already delivered %s",cachename);
 			result = 1;
 			break;
-
 		case VINE_CACHE_TRANSFER:
-			debug(D_VINE,"cache: transferring %s to %s",f->source,cachename);
 			result = do_transfer(c,f->source,cache_path,&error_message);
 			break;
 
 		case VINE_CACHE_MINI_TASK:
-			debug(D_VINE,"cache: creating %s via mini task",cachename);
-			result = do_mini_task(c,f->mini_task,manager,&error_message);
+			result = do_mini_task(c,f,&error_message);
 			break;
 	}
```

Comment:
Nothing is done with the error message here, at the very least send it to the debug stream.